In [327]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [328]:
data = pd.read_csv('retail_store_sales.csv')

# Retail Store Data Cleaning

The main purpose of this project is to perform data cleaning for a dataset that contains a retail store transaction. The dataset is from Kaggle ([source]('https://www.kaggle.com/datasets/ahmedmohamed2003/retail-store-sales-dirty-for-data-cleaning/data')).

The data is synthetic with 11 columns and 12,575 rows. The description of each column is:

- Transaction ID: A unique identifier for each transaction. Always present and unique.
- Customer ID: A unique identifier for each customer. 25 unique customers.
- Category: The category of the purchased item.
- Item: The name of the purchased item.
- Price Per Unit: The static price of a single unit of the item.
- Quantity: The quantity of the item purchased.
- Total Spent: The total amount spent on the transaction.
- Payment Method: The method of payment used.
- Location: The location where the transaction occurred.
- Transaction Date: The date of the transaction. Always present and valid.
- Discount Applied: Indicates if a discount was applied to the transaction.


In [329]:
data.head()

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   

   Quantity  Total Spent  Payment Method Location Transaction Date  \
0      10.0        185.0  Digital Wallet   Online       2024-04-08   
1       9.0        261.0  Digital Wallet   Online       2023-07-23   
2       2.0         43.0     Credit Card   Online       2022-10-05   
3       9.0        247.5     Credit Card   Online       2022-05-07   
4       7.0         87.5  Digital Wallet   Online       2022-10-02   

  Discount Applied  
0             True  
1             True  
2            False  
3              NaN  
4            False

In [330]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              11362 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Transaction Date  12575 non-null  object 
 10  Discount Applied  8376 non-null   object 
dtypes: float64(3), object(8)
memory usage: 1.1+ MB


The dataset has 12575 rows however, a lot of the columns contain null: Item, Price Per Unit, Quantity, Total Spent, Discount Applied. I will first try to fill these null data and only remove cells if the process affects the data integrity.

I will also change the data type of the Transaction Date columns to datetime64 and check for any duplicates and anomalies.

### 1. Checking for duplicates and changing the data type

In [331]:
data = data.drop_duplicates()
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Transaction ID    12575 non-null  object 
 1   Customer ID       12575 non-null  object 
 2   Category          12575 non-null  object 
 3   Item              11362 non-null  object 
 4   Price Per Unit    11966 non-null  float64
 5   Quantity          11971 non-null  float64
 6   Total Spent       11971 non-null  float64
 7   Payment Method    12575 non-null  object 
 8   Location          12575 non-null  object 
 9   Transaction Date  12575 non-null  object 
 10  Discount Applied  8376 non-null   object 
dtypes: float64(3), object(8)
memory usage: 1.1+ MB


The dataset doesn't have any duplicates as we can see that it still has 12575 rows.

In [332]:
data['Transaction Date'] = pd.to_datetime(data['Transaction Date'])
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    12575 non-null  object        
 1   Customer ID       12575 non-null  object        
 2   Category          12575 non-null  object        
 3   Item              11362 non-null  object        
 4   Price Per Unit    11966 non-null  float64       
 5   Quantity          11971 non-null  float64       
 6   Total Spent       11971 non-null  float64       
 7   Payment Method    12575 non-null  object        
 8   Location          12575 non-null  object        
 9   Transaction Date  12575 non-null  datetime64[ns]
 10  Discount Applied  8376 non-null   object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 1.1+ MB


### 2. Fill null values

In [333]:
data[data['Price Per Unit'].isnull()]

Transaction ID Customer ID                            Category Item  \
5        TXN_7482416     CUST_09                          Patisserie  NaN   
11       TXN_5422631     CUST_09                       Milk Products  NaN   
17       TXN_9634894     CUST_15                       Milk Products  NaN   
21       TXN_8685338     CUST_15                       Milk Products  NaN   
32       TXN_1543244     CUST_20                                Food  NaN   
...              ...         ...                                 ...  ...   
12435    TXN_5147764     CUST_01                           Furniture  NaN   
12457    TXN_1352194     CUST_17       Electric household essentials  NaN   
12477    TXN_5625684     CUST_22  Computers and electric accessories  NaN   
12491    TXN_7894525     CUST_23                            Butchers  NaN   
12511    TXN_5715500     CUST_17                                Food  NaN   

       Price Per Unit  Quantity  Total Spent  Payment Method  Location  \
5                 NaN      10.0        200.0     Credit Card    Online   
11                NaN       8.0         52.0  Digital Wallet  In-store   
17                NaN      10.0        275.0  Digital Wallet    Online   
21                NaN       3.0        105.0     Credit Card  In-store   
32                NaN       8.0        196.0     Credit Card    Online   
...               ...       ...          ...             ...       ...   
12435             NaN       5.0         32.5     Credit Card    Online   
12457             NaN       4.0         86.0     Credit Card    Online   
12477             NaN       4.0         80.0            Cash  In-store   
12491             NaN       1.0         26.0     Credit Card    Online   
12511             NaN       3.0        100.5  Digital Wallet    Online   

      Transaction Date Discount Applied  
5           2023-11-30              NaN  
11          2025-01-12             True  
17          2022-04-17              NaN  
21          2023-10-29              NaN  
32          2024-10-25             True  
...                ...              ...  
12435       2023-09-06             True  
12457       2023-02-26              NaN  
12477       2022-11-09             True  
12491       2023-01-31             True  
12511       2024-06-05            False  

[609 rows x 11 columns]

We can see that a lot of the prices are missing, however, we can use the quantity and total spent variables to fill in the missing price values using the formula Price Per Unit = Total Spent/Quantity.


However, we must first check if both Total Spent and Quantity values are available for these null Price Per Unit cells or not. 

In [334]:
data[data['Price Per Unit'].isnull()].info()

<class 'pandas.core.frame.DataFrame'>
Index: 609 entries, 5 to 12511
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    609 non-null    object        
 1   Customer ID       609 non-null    object        
 2   Category          609 non-null    object        
 3   Item              0 non-null      object        
 4   Price Per Unit    0 non-null      float64       
 5   Quantity          609 non-null    float64       
 6   Total Spent       609 non-null    float64       
 7   Payment Method    609 non-null    object        
 8   Location          609 non-null    object        
 9   Transaction Date  609 non-null    datetime64[ns]
 10  Discount Applied  404 non-null    object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 57.1+ KB


Both Quantity and Total Spent have 609 values (matches the number of null cells in Price Per Unit) so no further adjustment is required.

In [335]:
data['Price Per Unit'] = data['Price Per Unit'].fillna(data['Total Spent']/data['Quantity'])
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    12575 non-null  object        
 1   Customer ID       12575 non-null  object        
 2   Category          12575 non-null  object        
 3   Item              11362 non-null  object        
 4   Price Per Unit    12575 non-null  float64       
 5   Quantity          11971 non-null  float64       
 6   Total Spent       11971 non-null  float64       
 7   Payment Method    12575 non-null  object        
 8   Location          12575 non-null  object        
 9   Transaction Date  12575 non-null  datetime64[ns]
 10  Discount Applied  8376 non-null   object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 1.1+ MB


We have filled in all the null values for the Price Per Unit. 

I will next work on filling the null values for Total Spent and Quantity.

In [336]:
data[data['Total Spent'].isnull()].info()

<class 'pandas.core.frame.DataFrame'>
Index: 604 entries, 7 to 12564
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    604 non-null    object        
 1   Customer ID       604 non-null    object        
 2   Category          604 non-null    object        
 3   Item              0 non-null      object        
 4   Price Per Unit    604 non-null    float64       
 5   Quantity          0 non-null      float64       
 6   Total Spent       0 non-null      float64       
 7   Payment Method    604 non-null    object        
 8   Location          604 non-null    object        
 9   Transaction Date  604 non-null    datetime64[ns]
 10  Discount Applied  393 non-null    object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 56.6+ KB


In [337]:
data[data['Quantity'].isnull()].info()

<class 'pandas.core.frame.DataFrame'>
Index: 604 entries, 7 to 12564
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    604 non-null    object        
 1   Customer ID       604 non-null    object        
 2   Category          604 non-null    object        
 3   Item              0 non-null      object        
 4   Price Per Unit    604 non-null    float64       
 5   Quantity          0 non-null      float64       
 6   Total Spent       0 non-null      float64       
 7   Payment Method    604 non-null    object        
 8   Location          604 non-null    object        
 9   Transaction Date  604 non-null    datetime64[ns]
 10  Discount Applied  393 non-null    object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 56.6+ KB


We can see that the number of null values for the two columns Quantity and Total Spent are the same meaning that we can't apply the formula method to fill out these values.

Removing these rows won't be wise as there are 604 missing values which is almost half the dataset (12575 rows). Hence my approach will be to fill out the Quantity column first using the buying pattern from the customers then fill out the Total Spent using the formular Total Spent = Price Per Unit x Quantity.

To fill the Quantity columns, I will make an assumption that the customers buy a similar quantity in every transaction for a product. 

Using the Customer ID, Category and Price Per Unit(Item will be more accurate but we can see that the Item columns are also null), I will calculate the mean Quantity they bought and use that to fill in the Quantity column. This method ensures that it reflects the most likely scenario (for example customer A buys 5 units of product B in most transactions).

In [338]:
qty_null = data[data['Quantity'].isnull()]
qty_null

Transaction ID Customer ID                       Category Item  \
7        TXN_1372952     CUST_21                      Furniture  NaN   
15       TXN_1809665     CUST_14                      Beverages  NaN   
19       TXN_4206593     CUST_01                      Furniture  NaN   
25       TXN_3481599     CUST_05                      Furniture  NaN   
34       TXN_1621497     CUST_06                     Patisserie  NaN   
...              ...         ...                            ...  ...   
12527    TXN_1069238     CUST_23                           Food  NaN   
12552    TXN_4823896     CUST_05                  Milk Products  NaN   
12556    TXN_4397672     CUST_04                      Beverages  NaN   
12562    TXN_7422454     CUST_07                       Butchers  NaN   
12564    TXN_2153066     CUST_17  Electric household essentials  NaN   

       Price Per Unit  Quantity  Total Spent  Payment Method  Location  \
7                33.5       NaN          NaN  Digital Wallet  In-store   
15               24.5       NaN          NaN     Credit Card  In-store   
19               35.0       NaN          NaN  Digital Wallet    Online   
25               39.5       NaN          NaN            Cash    Online   
34               23.0       NaN          NaN            Cash  In-store   
...               ...       ...          ...             ...       ...   
12527             5.0       NaN          NaN  Digital Wallet  In-store   
12552             8.0       NaN          NaN            Cash  In-store   
12556            41.0       NaN          NaN     Credit Card    Online   
12562            33.5       NaN          NaN            Cash    Online   
12564            29.0       NaN          NaN  Digital Wallet  In-store   

      Transaction Date Discount Applied  
7           2024-04-02             True  
15          2022-05-11              NaN  
19          2025-01-13            False  
25          2022-09-08            False  
34          2023-02-18              NaN  
...                ...              ...  
12527       2022-08-13            False  
12552       2022-07-21            False  
12556       2024-11-28             True  
12562       2023-04-15              NaN  
12564       2024-03-28            False  

[604 rows x 11 columns]

In [339]:
qty = data[data['Quantity'].notnull()]
qty

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0        TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1        TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2        TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3        TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4        TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   
...              ...         ...            ...           ...             ...   
12570    TXN_9347481     CUST_18     Patisserie   Item_23_PAT            38.0   
12571    TXN_4009414     CUST_03      Beverages    Item_2_BEV             6.5   
12572    TXN_5306010     CUST_11       Butchers    Item_7_BUT            14.0   
12573    TXN_5167298     CUST_04      Furniture    Item_7_FUR            14.0   
12574    TXN_2407494     CUST_23           Food   Item_9_FOOD            17.0   

       Quantity  Total Spent  Payment Method  Location Transaction Date  \
0          10.0        185.0  Digital Wallet    Online       2024-04-08   
1           9.0        261.0  Digital Wallet    Online       2023-07-23   
2           2.0         43.0     Credit Card    Online       2022-10-05   
3           9.0        247.5     Credit Card    Online       2022-05-07   
4           7.0         87.5  Digital Wallet    Online       2022-10-02   
...         ...          ...             ...       ...              ...   
12570       4.0        152.0     Credit Card  In-store       2023-09-03   
12571       9.0         58.5            Cash    Online       2022-08-12   
12572      10.0        140.0            Cash    Online       2024-08-24   
12573       6.0         84.0            Cash    Online       2023-12-30   
12574       3.0         51.0            Cash    Online       2022-08-06   

      Discount Applied  
0                 True  
1                 True  
2                False  
3                  NaN  
4                False  
...                ...  
12570              NaN  
12571            False  
12572              NaN  
12573             True  
12574              NaN  

[11971 rows x 11 columns]

In [340]:
qty = qty.groupby(['Customer ID', 'Category','Price Per Unit'])['Quantity'].mean().round(0).reset_index()
qty

Customer ID    Category  Price Per Unit  Quantity
0        CUST_01   Beverages             5.0       5.0
1        CUST_01   Beverages             6.5       7.0
2        CUST_01   Beverages             8.0       7.0
3        CUST_01   Beverages             9.5       5.0
4        CUST_01   Beverages            14.0       3.0
...          ...         ...             ...       ...
4100     CUST_25  Patisserie            32.0       5.0
4101     CUST_25  Patisserie            33.5       8.0
4102     CUST_25  Patisserie            38.0       6.0
4103     CUST_25  Patisserie            39.5       7.0
4104     CUST_25  Patisserie            41.0       4.0

[4105 rows x 4 columns]

I have rounded the quantity to a whole number as all the products are sold/bought by unit. This might cause a rounding error, which is a potential issue to be aware of, but acceptable given the unit-based nature of the transactions.

In [341]:
merge_qty = pd.merge(qty_null, qty, on =['Customer ID','Category','Price Per Unit'], how = 'left', suffixes=("_null","_fill"))
merge_qty

Transaction ID Customer ID                       Category Item  \
0      TXN_1372952     CUST_21                      Furniture  NaN   
1      TXN_1809665     CUST_14                      Beverages  NaN   
2      TXN_4206593     CUST_01                      Furniture  NaN   
3      TXN_3481599     CUST_05                      Furniture  NaN   
4      TXN_1621497     CUST_06                     Patisserie  NaN   
..             ...         ...                            ...  ...   
599    TXN_1069238     CUST_23                           Food  NaN   
600    TXN_4823896     CUST_05                  Milk Products  NaN   
601    TXN_4397672     CUST_04                      Beverages  NaN   
602    TXN_7422454     CUST_07                       Butchers  NaN   
603    TXN_2153066     CUST_17  Electric household essentials  NaN   

     Price Per Unit  Quantity_null  Total Spent  Payment Method  Location  \
0              33.5            NaN          NaN  Digital Wallet  In-store   
1              24.5            NaN          NaN     Credit Card  In-store   
2              35.0            NaN          NaN  Digital Wallet    Online   
3              39.5            NaN          NaN            Cash    Online   
4              23.0            NaN          NaN            Cash  In-store   
..              ...            ...          ...             ...       ...   
599             5.0            NaN          NaN  Digital Wallet  In-store   
600             8.0            NaN          NaN            Cash  In-store   
601            41.0            NaN          NaN     Credit Card    Online   
602            33.5            NaN          NaN            Cash    Online   
603            29.0            NaN          NaN  Digital Wallet  In-store   

    Transaction Date Discount Applied  Quantity_fill  
0         2024-04-02             True            4.0  
1         2022-05-11              NaN            7.0  
2         2025-01-13            False            4.0  
3         2022-09-08            False            5.0  
4         2023-02-18              NaN            NaN  
..               ...              ...            ...  
599       2022-08-13            False            7.0  
600       2022-07-21            False            5.0  
601       2024-11-28             True            5.0  
602       2023-04-15              NaN            6.0  
603       2024-03-28            False            7.0  

[604 rows x 12 columns]

In [342]:
merge_qty.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 604 entries, 0 to 603
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    604 non-null    object        
 1   Customer ID       604 non-null    object        
 2   Category          604 non-null    object        
 3   Item              0 non-null      object        
 4   Price Per Unit    604 non-null    float64       
 5   Quantity_null     0 non-null      float64       
 6   Total Spent       0 non-null      float64       
 7   Payment Method    604 non-null    object        
 8   Location          604 non-null    object        
 9   Transaction Date  604 non-null    datetime64[ns]
 10  Discount Applied  393 non-null    object        
 11  Quantity_fill     560 non-null    float64       
dtypes: datetime64[ns](1), float64(4), object(7)
memory usage: 56.8+ KB


We're able to match 560 rows with exactly the same Customer ID, Category, and Price Per Unit. I will this in first and then assess the remaining null cells.

In [343]:
merge_qty['Quantity_null'] = merge_qty['Quantity_fill']
merge_qty = merge_qty.drop(columns='Quantity_fill')
merge_qty = merge_qty.rename(columns={'Quantity_null':"Quantity"})
merge_qty

Transaction ID Customer ID                       Category Item  \
0      TXN_1372952     CUST_21                      Furniture  NaN   
1      TXN_1809665     CUST_14                      Beverages  NaN   
2      TXN_4206593     CUST_01                      Furniture  NaN   
3      TXN_3481599     CUST_05                      Furniture  NaN   
4      TXN_1621497     CUST_06                     Patisserie  NaN   
..             ...         ...                            ...  ...   
599    TXN_1069238     CUST_23                           Food  NaN   
600    TXN_4823896     CUST_05                  Milk Products  NaN   
601    TXN_4397672     CUST_04                      Beverages  NaN   
602    TXN_7422454     CUST_07                       Butchers  NaN   
603    TXN_2153066     CUST_17  Electric household essentials  NaN   

     Price Per Unit  Quantity  Total Spent  Payment Method  Location  \
0              33.5       4.0          NaN  Digital Wallet  In-store   
1              24.5       7.0          NaN     Credit Card  In-store   
2              35.0       4.0          NaN  Digital Wallet    Online   
3              39.5       5.0          NaN            Cash    Online   
4              23.0       NaN          NaN            Cash  In-store   
..              ...       ...          ...             ...       ...   
599             5.0       7.0          NaN  Digital Wallet  In-store   
600             8.0       5.0          NaN            Cash  In-store   
601            41.0       5.0          NaN     Credit Card    Online   
602            33.5       6.0          NaN            Cash    Online   
603            29.0       7.0          NaN  Digital Wallet  In-store   

    Transaction Date Discount Applied  
0         2024-04-02             True  
1         2022-05-11              NaN  
2         2025-01-13            False  
3         2022-09-08            False  
4         2023-02-18              NaN  
..               ...              ...  
599       2022-08-13            False  
600       2022-07-21            False  
601       2024-11-28             True  
602       2023-04-15              NaN  
603       2024-03-28            False  

[604 rows x 11 columns]

In [344]:
merge_qty[merge_qty['Quantity'].isnull()].head()

Transaction ID Customer ID                       Category Item  \
4     TXN_1621497     CUST_06                     Patisserie  NaN   
27    TXN_3171192     CUST_13                           Food  NaN   
30    TXN_5034035     CUST_08                      Furniture  NaN   
37    TXN_7705441     CUST_25                  Milk Products  NaN   
39    TXN_1909279     CUST_10  Electric household essentials  NaN   

    Price Per Unit  Quantity  Total Spent  Payment Method  Location  \
4             23.0       NaN          NaN            Cash  In-store   
27            26.0       NaN          NaN            Cash    Online   
30             9.5       NaN          NaN     Credit Card    Online   
37            26.0       NaN          NaN  Digital Wallet    Online   
39            24.5       NaN          NaN     Credit Card    Online   

   Transaction Date Discount Applied  
4        2023-02-18              NaN  
27       2024-01-22             True  
30       2024-08-25            False  
37       2022-02-27            False  
39       2024-06-12             True

In [345]:
merge_qty[merge_qty['Quantity'].isnull()].info()

<class 'pandas.core.frame.DataFrame'>
Index: 44 entries, 4 to 594
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    44 non-null     object        
 1   Customer ID       44 non-null     object        
 2   Category          44 non-null     object        
 3   Item              0 non-null      object        
 4   Price Per Unit    44 non-null     float64       
 5   Quantity          0 non-null      float64       
 6   Total Spent       0 non-null      float64       
 7   Payment Method    44 non-null     object        
 8   Location          44 non-null     object        
 9   Transaction Date  44 non-null     datetime64[ns]
 10  Discount Applied  30 non-null     object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 4.1+ KB


In [346]:
qty[(qty['Customer ID']=='CUST_06') & (qty['Category']=='Patisserie')]

Customer ID    Category  Price Per Unit  Quantity
985      CUST_06  Patisserie             5.0       6.0
986      CUST_06  Patisserie             6.5       8.0
987      CUST_06  Patisserie             8.0       2.0
988      CUST_06  Patisserie             9.5       3.0
989      CUST_06  Patisserie            12.5       5.0
990      CUST_06  Patisserie            14.0       6.0
991      CUST_06  Patisserie            15.5       7.0
992      CUST_06  Patisserie            17.0       6.0
993      CUST_06  Patisserie            18.5       7.0
994      CUST_06  Patisserie            20.0       8.0
995      CUST_06  Patisserie            21.5       6.0
996      CUST_06  Patisserie            24.5       4.0
997      CUST_06  Patisserie            27.5       3.0
998      CUST_06  Patisserie            29.0       6.0
999      CUST_06  Patisserie            30.5       2.0
1000     CUST_06  Patisserie            33.5       2.0
1001     CUST_06  Patisserie            36.5       7.0
1002     CUST_06  Patisserie            38.0       8.0
1003     CUST_06  Patisserie            39.5       5.0

In [347]:
qty[(qty['Customer ID']=='CUST_13') & (qty['Category']=='Food')]

Customer ID Category  Price Per Unit  Quantity
2053     CUST_13     Food             5.0       4.0
2054     CUST_13     Food             8.0       7.0
2055     CUST_13     Food             9.5       4.0
2056     CUST_13     Food            11.0       5.0
2057     CUST_13     Food            12.5       6.0
2058     CUST_13     Food            17.0       6.0
2059     CUST_13     Food            18.5       8.0
2060     CUST_13     Food            20.0       5.0
2061     CUST_13     Food            21.5       8.0
2062     CUST_13     Food            23.0       6.0
2063     CUST_13     Food            24.5       6.0
2064     CUST_13     Food            27.5       2.0
2065     CUST_13     Food            29.0       3.0
2066     CUST_13     Food            30.5       7.0
2067     CUST_13     Food            33.5       6.0
2068     CUST_13     Food            36.5       8.0
2069     CUST_13     Food            38.0       5.0
2070     CUST_13     Food            39.5       2.0
2071     CUST_13     Food            41.0      10.0

In [348]:
((604 - 560)/12575)*100

0.3499005964214712

We can see that without knowing the Item, the Quantity fluctuates a lot. One way to fill them in is to use the mean or median, however, it will violate the assumption I made (customers buy the same quantity for the same product in each transaction) and I don't think it will give a very accurate result. Therefore, it would be best to remove these null rows (44 rows) as they only take up 0.35% of the dataset.

In [349]:
merge_qty = merge_qty[merge_qty['Quantity'].notnull()]
merge_qty

Transaction ID Customer ID                       Category Item  \
0      TXN_1372952     CUST_21                      Furniture  NaN   
1      TXN_1809665     CUST_14                      Beverages  NaN   
2      TXN_4206593     CUST_01                      Furniture  NaN   
3      TXN_3481599     CUST_05                      Furniture  NaN   
5      TXN_4532214     CUST_25                     Patisserie  NaN   
..             ...         ...                            ...  ...   
599    TXN_1069238     CUST_23                           Food  NaN   
600    TXN_4823896     CUST_05                  Milk Products  NaN   
601    TXN_4397672     CUST_04                      Beverages  NaN   
602    TXN_7422454     CUST_07                       Butchers  NaN   
603    TXN_2153066     CUST_17  Electric household essentials  NaN   

     Price Per Unit  Quantity  Total Spent  Payment Method  Location  \
0              33.5       4.0          NaN  Digital Wallet  In-store   
1              24.5       7.0          NaN     Credit Card  In-store   
2              35.0       4.0          NaN  Digital Wallet    Online   
3              39.5       5.0          NaN            Cash    Online   
5              15.5       4.0          NaN     Credit Card  In-store   
..              ...       ...          ...             ...       ...   
599             5.0       7.0          NaN  Digital Wallet  In-store   
600             8.0       5.0          NaN            Cash  In-store   
601            41.0       5.0          NaN     Credit Card    Online   
602            33.5       6.0          NaN            Cash    Online   
603            29.0       7.0          NaN  Digital Wallet  In-store   

    Transaction Date Discount Applied  
0         2024-04-02             True  
1         2022-05-11              NaN  
2         2025-01-13            False  
3         2022-09-08            False  
5         2024-03-25            False  
..               ...              ...  
599       2022-08-13            False  
600       2022-07-21            False  
601       2024-11-28             True  
602       2023-04-15              NaN  
603       2024-03-28            False  

[560 rows x 11 columns]

In [350]:
data_qty = pd.merge(data,merge_qty,on = ['Transaction ID','Customer ID','Category','Item','Price Per Unit','Total Spent','Payment Method','Location','Transaction Date','Discount Applied'], how = 'left', suffixes=("_null","_fill"))
data_qty.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    12575 non-null  object        
 1   Customer ID       12575 non-null  object        
 2   Category          12575 non-null  object        
 3   Item              11362 non-null  object        
 4   Price Per Unit    12575 non-null  float64       
 5   Quantity_null     11971 non-null  float64       
 6   Total Spent       11971 non-null  float64       
 7   Payment Method    12575 non-null  object        
 8   Location          12575 non-null  object        
 9   Transaction Date  12575 non-null  datetime64[ns]
 10  Discount Applied  8376 non-null   object        
 11  Quantity_fill     560 non-null    float64       
dtypes: datetime64[ns](1), float64(4), object(7)
memory usage: 1.2+ MB


In [351]:
data_qty['Quantity_null'] = data_qty['Quantity_null'].combine_first(data_qty['Quantity_fill'])
data_qty

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0        TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1        TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2        TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3        TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4        TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   
...              ...         ...            ...           ...             ...   
12570    TXN_9347481     CUST_18     Patisserie   Item_23_PAT            38.0   
12571    TXN_4009414     CUST_03      Beverages    Item_2_BEV             6.5   
12572    TXN_5306010     CUST_11       Butchers    Item_7_BUT            14.0   
12573    TXN_5167298     CUST_04      Furniture    Item_7_FUR            14.0   
12574    TXN_2407494     CUST_23           Food   Item_9_FOOD            17.0   

       Quantity_null  Total Spent  Payment Method  Location Transaction Date  \
0               10.0        185.0  Digital Wallet    Online       2024-04-08   
1                9.0        261.0  Digital Wallet    Online       2023-07-23   
2                2.0         43.0     Credit Card    Online       2022-10-05   
3                9.0        247.5     Credit Card    Online       2022-05-07   
4                7.0         87.5  Digital Wallet    Online       2022-10-02   
...              ...          ...             ...       ...              ...   
12570            4.0        152.0     Credit Card  In-store       2023-09-03   
12571            9.0         58.5            Cash    Online       2022-08-12   
12572           10.0        140.0            Cash    Online       2024-08-24   
12573            6.0         84.0            Cash    Online       2023-12-30   
12574            3.0         51.0            Cash    Online       2022-08-06   

      Discount Applied  Quantity_fill  
0                 True            NaN  
1                 True            NaN  
2                False            NaN  
3                  NaN            NaN  
4                False            NaN  
...                ...            ...  
12570              NaN            NaN  
12571            False            NaN  
12572              NaN            NaN  
12573             True            NaN  
12574              NaN            NaN  

[12575 rows x 12 columns]

In [352]:
data_qty.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12575 entries, 0 to 12574
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    12575 non-null  object        
 1   Customer ID       12575 non-null  object        
 2   Category          12575 non-null  object        
 3   Item              11362 non-null  object        
 4   Price Per Unit    12575 non-null  float64       
 5   Quantity_null     12531 non-null  float64       
 6   Total Spent       11971 non-null  float64       
 7   Payment Method    12575 non-null  object        
 8   Location          12575 non-null  object        
 9   Transaction Date  12575 non-null  datetime64[ns]
 10  Discount Applied  8376 non-null   object        
 11  Quantity_fill     560 non-null    float64       
dtypes: datetime64[ns](1), float64(4), object(7)
memory usage: 1.2+ MB


In [353]:
data_qty = data_qty.drop(columns=['Quantity_fill'])
data_qty = data_qty.rename(columns={'Quantity_null':'Quantity'})
data_qty = data_qty[data_qty['Quantity'].notnull()]
data_qty

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0        TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1        TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2        TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3        TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4        TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   
...              ...         ...            ...           ...             ...   
12570    TXN_9347481     CUST_18     Patisserie   Item_23_PAT            38.0   
12571    TXN_4009414     CUST_03      Beverages    Item_2_BEV             6.5   
12572    TXN_5306010     CUST_11       Butchers    Item_7_BUT            14.0   
12573    TXN_5167298     CUST_04      Furniture    Item_7_FUR            14.0   
12574    TXN_2407494     CUST_23           Food   Item_9_FOOD            17.0   

       Quantity  Total Spent  Payment Method  Location Transaction Date  \
0          10.0        185.0  Digital Wallet    Online       2024-04-08   
1           9.0        261.0  Digital Wallet    Online       2023-07-23   
2           2.0         43.0     Credit Card    Online       2022-10-05   
3           9.0        247.5     Credit Card    Online       2022-05-07   
4           7.0         87.5  Digital Wallet    Online       2022-10-02   
...         ...          ...             ...       ...              ...   
12570       4.0        152.0     Credit Card  In-store       2023-09-03   
12571       9.0         58.5            Cash    Online       2022-08-12   
12572      10.0        140.0            Cash    Online       2024-08-24   
12573       6.0         84.0            Cash    Online       2023-12-30   
12574       3.0         51.0            Cash    Online       2022-08-06   

      Discount Applied  
0                 True  
1                 True  
2                False  
3                  NaN  
4                False  
...                ...  
12570              NaN  
12571            False  
12572              NaN  
12573             True  
12574              NaN  

[12531 rows x 11 columns]

In [354]:
data_qty.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12531 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    12531 non-null  object        
 1   Customer ID       12531 non-null  object        
 2   Category          12531 non-null  object        
 3   Item              11362 non-null  object        
 4   Price Per Unit    12531 non-null  float64       
 5   Quantity          12531 non-null  float64       
 6   Total Spent       11971 non-null  float64       
 7   Payment Method    12531 non-null  object        
 8   Location          12531 non-null  object        
 9   Transaction Date  12531 non-null  datetime64[ns]
 10  Discount Applied  8346 non-null   object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 1.1+ MB


Now that we have removed the 44 rows mentioned above and filled in the Quantity, the next time is to fill in the Total Spent, Item and Discount Applied.

In [355]:
data_qty['Total Spent'] = data_qty['Total Spent'].fillna(data_qty['Price Per Unit']*data_qty['Quantity'])
data_qty.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12531 entries, 0 to 12574
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    12531 non-null  object        
 1   Customer ID       12531 non-null  object        
 2   Category          12531 non-null  object        
 3   Item              11362 non-null  object        
 4   Price Per Unit    12531 non-null  float64       
 5   Quantity          12531 non-null  float64       
 6   Total Spent       12531 non-null  float64       
 7   Payment Method    12531 non-null  object        
 8   Location          12531 non-null  object        
 9   Transaction Date  12531 non-null  datetime64[ns]
 10  Discount Applied  8346 non-null   object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 1.1+ MB


In [356]:
item = data_qty.groupby(['Category','Item','Price Per Unit','Discount Applied'])['Transaction ID'].count().reset_index()
item

Category         Item  Price Per Unit  Discount Applied  Transaction ID
0     Beverages  Item_10_BEV            18.5             False               8
1     Beverages  Item_10_BEV            18.5              True              10
2     Beverages  Item_11_BEV            20.0             False              19
3     Beverages  Item_11_BEV            20.0              True              19
4     Beverages  Item_12_BEV            21.5             False              35
..          ...          ...             ...               ...             ...
394  Patisserie   Item_7_PAT            14.0              True              17
395  Patisserie   Item_8_PAT            15.5             False              27
396  Patisserie   Item_8_PAT            15.5              True              25
397  Patisserie   Item_9_PAT            17.0             False              20
398  Patisserie   Item_9_PAT            17.0              True              20

[399 rows x 5 columns]

In [357]:
item_null = data_qty[data_qty['Item'].isnull()]
item_null

Transaction ID Customer ID                       Category Item  \
5        TXN_7482416     CUST_09                     Patisserie  NaN   
7        TXN_1372952     CUST_21                      Furniture  NaN   
11       TXN_5422631     CUST_09                  Milk Products  NaN   
15       TXN_1809665     CUST_14                      Beverages  NaN   
17       TXN_9634894     CUST_15                  Milk Products  NaN   
...              ...         ...                            ...  ...   
12527    TXN_1069238     CUST_23                           Food  NaN   
12552    TXN_4823896     CUST_05                  Milk Products  NaN   
12556    TXN_4397672     CUST_04                      Beverages  NaN   
12562    TXN_7422454     CUST_07                       Butchers  NaN   
12564    TXN_2153066     CUST_17  Electric household essentials  NaN   

       Price Per Unit  Quantity  Total Spent  Payment Method  Location  \
5                20.0      10.0        200.0     Credit Card    Online   
7                33.5       4.0        134.0  Digital Wallet  In-store   
11                6.5       8.0         52.0  Digital Wallet  In-store   
15               24.5       7.0        171.5     Credit Card  In-store   
17               27.5      10.0        275.0  Digital Wallet    Online   
...               ...       ...          ...             ...       ...   
12527             5.0       7.0         35.0  Digital Wallet  In-store   
12552             8.0       5.0         40.0            Cash  In-store   
12556            41.0       5.0        205.0     Credit Card    Online   
12562            33.5       6.0        201.0            Cash    Online   
12564            29.0       7.0        203.0  Digital Wallet  In-store   

      Transaction Date Discount Applied  
5           2023-11-30              NaN  
7           2024-04-02             True  
11          2025-01-12             True  
15          2022-05-11              NaN  
17          2022-04-17              NaN  
...                ...              ...  
12527       2022-08-13            False  
12552       2022-07-21            False  
12556       2024-11-28             True  
12562       2023-04-15              NaN  
12564       2024-03-28            False  

[1169 rows x 11 columns]

In [358]:
merge_item = pd.merge(item_null,item,on=['Category','Price Per Unit','Discount Applied'], how='left', suffixes =('_null','_fill'))
merge_item

Transaction ID_null Customer ID                       Category Item_null  \
0            TXN_7482416     CUST_09                     Patisserie       NaN   
1            TXN_1372952     CUST_21                      Furniture       NaN   
2            TXN_5422631     CUST_09                  Milk Products       NaN   
3            TXN_1809665     CUST_14                      Beverages       NaN   
4            TXN_9634894     CUST_15                  Milk Products       NaN   
...                  ...         ...                            ...       ...   
1164         TXN_1069238     CUST_23                           Food       NaN   
1165         TXN_4823896     CUST_05                  Milk Products       NaN   
1166         TXN_4397672     CUST_04                      Beverages       NaN   
1167         TXN_7422454     CUST_07                       Butchers       NaN   
1168         TXN_2153066     CUST_17  Electric household essentials       NaN   

      Price Per Unit  Quantity  Total Spent  Payment Method  Location  \
0               20.0      10.0        200.0     Credit Card    Online   
1               33.5       4.0        134.0  Digital Wallet  In-store   
2                6.5       8.0         52.0  Digital Wallet  In-store   
3               24.5       7.0        171.5     Credit Card  In-store   
4               27.5      10.0        275.0  Digital Wallet    Online   
...              ...       ...          ...             ...       ...   
1164             5.0       7.0         35.0  Digital Wallet  In-store   
1165             8.0       5.0         40.0            Cash  In-store   
1166            41.0       5.0        205.0     Credit Card    Online   
1167            33.5       6.0        201.0            Cash    Online   
1168            29.0       7.0        203.0  Digital Wallet  In-store   

     Transaction Date Discount Applied    Item_fill  Transaction ID_fill  
0          2023-11-30              NaN          NaN                  NaN  
1          2024-04-02             True  Item_20_FUR                 21.0  
2          2025-01-12             True  Item_2_MILK                 17.0  
3          2022-05-11              NaN          NaN                  NaN  
4          2022-04-17              NaN          NaN                  NaN  
...               ...              ...          ...                  ...  
1164       2022-08-13            False  Item_1_FOOD                  6.0  
1165       2022-07-21            False  Item_3_MILK                 30.0  
1166       2024-11-28             True  Item_25_BEV                 30.0  
1167       2023-04-15              NaN          NaN                  NaN  
1168       2024-03-28            False  Item_17_EHE                 17.0  

[1169 rows x 13 columns]

In [359]:
merge_item.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1169 entries, 0 to 1168
Data columns (total 13 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Transaction ID_null  1169 non-null   object        
 1   Customer ID          1169 non-null   object        
 2   Category             1169 non-null   object        
 3   Item_null            0 non-null      object        
 4   Price Per Unit       1169 non-null   float64       
 5   Quantity             1169 non-null   float64       
 6   Total Spent          1169 non-null   float64       
 7   Payment Method       1169 non-null   object        
 8   Location             1169 non-null   object        
 9   Transaction Date     1169 non-null   datetime64[ns]
 10  Discount Applied     767 non-null    object        
 11  Item_fill            767 non-null    object        
 12  Transaction ID_fill  767 non-null    float64       
dtypes: datetime64[ns](1), float64(4),

We are able to fill 767 cells for the Item column with this method out of 1169 null cells.

In [360]:
merge_item['Item_null'] = merge_item['Item_fill']
merge_item = merge_item.drop(columns=['Item_fill','Transaction ID_fill'])
merge_item = merge_item.rename(columns={"Item_null":"Item",'Transaction ID_null':'Transaction ID'})
merge_item

Transaction ID Customer ID                       Category         Item  \
0       TXN_7482416     CUST_09                     Patisserie          NaN   
1       TXN_1372952     CUST_21                      Furniture  Item_20_FUR   
2       TXN_5422631     CUST_09                  Milk Products  Item_2_MILK   
3       TXN_1809665     CUST_14                      Beverages          NaN   
4       TXN_9634894     CUST_15                  Milk Products          NaN   
...             ...         ...                            ...          ...   
1164    TXN_1069238     CUST_23                           Food  Item_1_FOOD   
1165    TXN_4823896     CUST_05                  Milk Products  Item_3_MILK   
1166    TXN_4397672     CUST_04                      Beverages  Item_25_BEV   
1167    TXN_7422454     CUST_07                       Butchers          NaN   
1168    TXN_2153066     CUST_17  Electric household essentials  Item_17_EHE   

      Price Per Unit  Quantity  Total Spent  Payment Method  Location  \
0               20.0      10.0        200.0     Credit Card    Online   
1               33.5       4.0        134.0  Digital Wallet  In-store   
2                6.5       8.0         52.0  Digital Wallet  In-store   
3               24.5       7.0        171.5     Credit Card  In-store   
4               27.5      10.0        275.0  Digital Wallet    Online   
...              ...       ...          ...             ...       ...   
1164             5.0       7.0         35.0  Digital Wallet  In-store   
1165             8.0       5.0         40.0            Cash  In-store   
1166            41.0       5.0        205.0     Credit Card    Online   
1167            33.5       6.0        201.0            Cash    Online   
1168            29.0       7.0        203.0  Digital Wallet  In-store   

     Transaction Date Discount Applied  
0          2023-11-30              NaN  
1          2024-04-02             True  
2          2025-01-12             True  
3          2022-05-11              NaN  
4          2022-04-17              NaN  
...               ...              ...  
1164       2022-08-13            False  
1165       2022-07-21            False  
1166       2024-11-28             True  
1167       2023-04-15              NaN  
1168       2024-03-28            False  

[1169 rows x 11 columns]

For the rest of the null values in Item, I will replace them with 'Unknown'.

In [361]:
merge_item['Item']=merge_item['Item'].fillna('Unknown')
merge_item

Transaction ID Customer ID                       Category         Item  \
0       TXN_7482416     CUST_09                     Patisserie      Unknown   
1       TXN_1372952     CUST_21                      Furniture  Item_20_FUR   
2       TXN_5422631     CUST_09                  Milk Products  Item_2_MILK   
3       TXN_1809665     CUST_14                      Beverages      Unknown   
4       TXN_9634894     CUST_15                  Milk Products      Unknown   
...             ...         ...                            ...          ...   
1164    TXN_1069238     CUST_23                           Food  Item_1_FOOD   
1165    TXN_4823896     CUST_05                  Milk Products  Item_3_MILK   
1166    TXN_4397672     CUST_04                      Beverages  Item_25_BEV   
1167    TXN_7422454     CUST_07                       Butchers      Unknown   
1168    TXN_2153066     CUST_17  Electric household essentials  Item_17_EHE   

      Price Per Unit  Quantity  Total Spent  Payment Method  Location  \
0               20.0      10.0        200.0     Credit Card    Online   
1               33.5       4.0        134.0  Digital Wallet  In-store   
2                6.5       8.0         52.0  Digital Wallet  In-store   
3               24.5       7.0        171.5     Credit Card  In-store   
4               27.5      10.0        275.0  Digital Wallet    Online   
...              ...       ...          ...             ...       ...   
1164             5.0       7.0         35.0  Digital Wallet  In-store   
1165             8.0       5.0         40.0            Cash  In-store   
1166            41.0       5.0        205.0     Credit Card    Online   
1167            33.5       6.0        201.0            Cash    Online   
1168            29.0       7.0        203.0  Digital Wallet  In-store   

     Transaction Date Discount Applied  
0          2023-11-30              NaN  
1          2024-04-02             True  
2          2025-01-12             True  
3          2022-05-11              NaN  
4          2022-04-17              NaN  
...               ...              ...  
1164       2022-08-13            False  
1165       2022-07-21            False  
1166       2024-11-28             True  
1167       2023-04-15              NaN  
1168       2024-03-28            False  

[1169 rows x 11 columns]

In [362]:
merge_item.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1169 entries, 0 to 1168
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    1169 non-null   object        
 1   Customer ID       1169 non-null   object        
 2   Category          1169 non-null   object        
 3   Item              1169 non-null   object        
 4   Price Per Unit    1169 non-null   float64       
 5   Quantity          1169 non-null   float64       
 6   Total Spent       1169 non-null   float64       
 7   Payment Method    1169 non-null   object        
 8   Location          1169 non-null   object        
 9   Transaction Date  1169 non-null   datetime64[ns]
 10  Discount Applied  767 non-null    object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 100.6+ KB


In [363]:
data_item = pd.merge(data_qty,merge_item,on=['Transaction ID','Customer ID','Category','Price Per Unit','Quantity','Total Spent','Payment Method','Location','Transaction Date','Discount Applied'],how='left',suffixes=('_null','_fill'))
data_item

Transaction ID Customer ID       Category     Item_null  Price Per Unit  \
0        TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1        TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2        TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3        TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4        TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   
...              ...         ...            ...           ...             ...   
12526    TXN_9347481     CUST_18     Patisserie   Item_23_PAT            38.0   
12527    TXN_4009414     CUST_03      Beverages    Item_2_BEV             6.5   
12528    TXN_5306010     CUST_11       Butchers    Item_7_BUT            14.0   
12529    TXN_5167298     CUST_04      Furniture    Item_7_FUR            14.0   
12530    TXN_2407494     CUST_23           Food   Item_9_FOOD            17.0   

       Quantity  Total Spent  Payment Method  Location Transaction Date  \
0          10.0        185.0  Digital Wallet    Online       2024-04-08   
1           9.0        261.0  Digital Wallet    Online       2023-07-23   
2           2.0         43.0     Credit Card    Online       2022-10-05   
3           9.0        247.5     Credit Card    Online       2022-05-07   
4           7.0         87.5  Digital Wallet    Online       2022-10-02   
...         ...          ...             ...       ...              ...   
12526       4.0        152.0     Credit Card  In-store       2023-09-03   
12527       9.0         58.5            Cash    Online       2022-08-12   
12528      10.0        140.0            Cash    Online       2024-08-24   
12529       6.0         84.0            Cash    Online       2023-12-30   
12530       3.0         51.0            Cash    Online       2022-08-06   

      Discount Applied Item_fill  
0                 True       NaN  
1                 True       NaN  
2                False       NaN  
3                  NaN       NaN  
4                False       NaN  
...                ...       ...  
12526              NaN       NaN  
12527            False       NaN  
12528              NaN       NaN  
12529             True       NaN  
12530              NaN       NaN  

[12531 rows x 12 columns]

In [364]:
data_item['Item_null'] = data_item['Item_null'].combine_first(data_item['Item_fill'])
data_item

Transaction ID Customer ID       Category     Item_null  Price Per Unit  \
0        TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1        TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2        TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3        TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4        TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   
...              ...         ...            ...           ...             ...   
12526    TXN_9347481     CUST_18     Patisserie   Item_23_PAT            38.0   
12527    TXN_4009414     CUST_03      Beverages    Item_2_BEV             6.5   
12528    TXN_5306010     CUST_11       Butchers    Item_7_BUT            14.0   
12529    TXN_5167298     CUST_04      Furniture    Item_7_FUR            14.0   
12530    TXN_2407494     CUST_23           Food   Item_9_FOOD            17.0   

       Quantity  Total Spent  Payment Method  Location Transaction Date  \
0          10.0        185.0  Digital Wallet    Online       2024-04-08   
1           9.0        261.0  Digital Wallet    Online       2023-07-23   
2           2.0         43.0     Credit Card    Online       2022-10-05   
3           9.0        247.5     Credit Card    Online       2022-05-07   
4           7.0         87.5  Digital Wallet    Online       2022-10-02   
...         ...          ...             ...       ...              ...   
12526       4.0        152.0     Credit Card  In-store       2023-09-03   
12527       9.0         58.5            Cash    Online       2022-08-12   
12528      10.0        140.0            Cash    Online       2024-08-24   
12529       6.0         84.0            Cash    Online       2023-12-30   
12530       3.0         51.0            Cash    Online       2022-08-06   

      Discount Applied Item_fill  
0                 True       NaN  
1                 True       NaN  
2                False       NaN  
3                  NaN       NaN  
4                False       NaN  
...                ...       ...  
12526              NaN       NaN  
12527            False       NaN  
12528              NaN       NaN  
12529             True       NaN  
12530              NaN       NaN  

[12531 rows x 12 columns]

In [365]:
data_item.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12531 entries, 0 to 12530
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    12531 non-null  object        
 1   Customer ID       12531 non-null  object        
 2   Category          12531 non-null  object        
 3   Item_null         12531 non-null  object        
 4   Price Per Unit    12531 non-null  float64       
 5   Quantity          12531 non-null  float64       
 6   Total Spent       12531 non-null  float64       
 7   Payment Method    12531 non-null  object        
 8   Location          12531 non-null  object        
 9   Transaction Date  12531 non-null  datetime64[ns]
 10  Discount Applied  8346 non-null   object        
 11  Item_fill         1169 non-null   object        
dtypes: datetime64[ns](1), float64(3), object(8)
memory usage: 1.1+ MB


In [366]:
data_item = data_item.rename(columns={'Item_null':'Item'})
data_item = data_item.drop(columns='Item_fill')
data_item

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0        TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1        TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2        TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3        TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4        TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   
...              ...         ...            ...           ...             ...   
12526    TXN_9347481     CUST_18     Patisserie   Item_23_PAT            38.0   
12527    TXN_4009414     CUST_03      Beverages    Item_2_BEV             6.5   
12528    TXN_5306010     CUST_11       Butchers    Item_7_BUT            14.0   
12529    TXN_5167298     CUST_04      Furniture    Item_7_FUR            14.0   
12530    TXN_2407494     CUST_23           Food   Item_9_FOOD            17.0   

       Quantity  Total Spent  Payment Method  Location Transaction Date  \
0          10.0        185.0  Digital Wallet    Online       2024-04-08   
1           9.0        261.0  Digital Wallet    Online       2023-07-23   
2           2.0         43.0     Credit Card    Online       2022-10-05   
3           9.0        247.5     Credit Card    Online       2022-05-07   
4           7.0         87.5  Digital Wallet    Online       2022-10-02   
...         ...          ...             ...       ...              ...   
12526       4.0        152.0     Credit Card  In-store       2023-09-03   
12527       9.0         58.5            Cash    Online       2022-08-12   
12528      10.0        140.0            Cash    Online       2024-08-24   
12529       6.0         84.0            Cash    Online       2023-12-30   
12530       3.0         51.0            Cash    Online       2022-08-06   

      Discount Applied  
0                 True  
1                 True  
2                False  
3                  NaN  
4                False  
...                ...  
12526              NaN  
12527            False  
12528              NaN  
12529             True  
12530              NaN  

[12531 rows x 11 columns]

The last column to fill is Discount Applied. I will again use the similar method above for this column.

In [367]:
disap = data_item.groupby(['Category','Item','Price Per Unit','Discount Applied'])['Transaction ID'].count().reset_index()
disap

Category         Item  Price Per Unit  Discount Applied  Transaction ID
0     Beverages  Item_10_BEV            18.5             False               8
1     Beverages  Item_10_BEV            18.5              True              10
2     Beverages  Item_11_BEV            20.0             False              23
3     Beverages  Item_11_BEV            20.0              True              20
4     Beverages  Item_12_BEV            21.5             False              40
..          ...          ...             ...               ...             ...
394  Patisserie   Item_7_PAT            14.0              True              18
395  Patisserie   Item_8_PAT            15.5             False              34
396  Patisserie   Item_8_PAT            15.5              True              28
397  Patisserie   Item_9_PAT            17.0             False              20
398  Patisserie   Item_9_PAT            17.0              True              22

[399 rows x 5 columns]

We encounter a problem here as we can see that the same product (same category, item code and price) has both options for Discount Applied (True and False). Therefore, we can't match the null values based on this table. Therefore, the only option I have left is to fill out the null values with Unknown. 

In [368]:
final_data = data_item
final_data['Discount Applied'] = final_data['Discount Applied'].fillna('Unkown')
final_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12531 entries, 0 to 12530
Data columns (total 11 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction ID    12531 non-null  object        
 1   Customer ID       12531 non-null  object        
 2   Category          12531 non-null  object        
 3   Item              12531 non-null  object        
 4   Price Per Unit    12531 non-null  float64       
 5   Quantity          12531 non-null  float64       
 6   Total Spent       12531 non-null  float64       
 7   Payment Method    12531 non-null  object        
 8   Location          12531 non-null  object        
 9   Transaction Date  12531 non-null  datetime64[ns]
 10  Discount Applied  12531 non-null  object        
dtypes: datetime64[ns](1), float64(3), object(7)
memory usage: 1.1+ MB


In [369]:
final_data.head()

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   

   Quantity  Total Spent  Payment Method Location Transaction Date  \
0      10.0        185.0  Digital Wallet   Online       2024-04-08   
1       9.0        261.0  Digital Wallet   Online       2023-07-23   
2       2.0         43.0     Credit Card   Online       2022-10-05   
3       9.0        247.5     Credit Card   Online       2022-05-07   
4       7.0         87.5  Digital Wallet   Online       2022-10-02   

  Discount Applied  
0             True  
1             True  
2            False  
3           Unkown  
4            False

I have finished filling out the all the null values. The final step of data cleaning is to look for any anomaly in the dataset.

### 3. Identifying anomaly

In [370]:
final_data.describe()

Price Per Unit      Quantity   Total Spent  \
count    12531.000000  12531.000000  12531.000000   
mean        23.370002      5.532599    129.599673   
min          5.000000      1.000000      5.000000   
25%         14.000000      3.000000     52.000000   
50%         23.000000      6.000000    108.500000   
75%         33.500000      8.000000    192.000000   
max         41.000000     10.000000    410.000000   
std         10.753255      2.823166     94.041832   

                    Transaction Date  
count                          12531  
mean   2023-07-12 18:40:04.596600320  
min              2022-01-01 00:00:00  
25%              2022-09-30 00:00:00  
50%              2023-07-13 00:00:00  
75%              2024-04-24 00:00:00  
max              2025-01-18 00:00:00  
std                              NaN

Based on the statistics of the columns Price Per Unit and Quantity, the min and max are reasonable. Hence, I don't think there are any anomalies in this cleaned dataset.

In [371]:
final_data.head()

Transaction ID Customer ID       Category          Item  Price Per Unit  \
0    TXN_6867343     CUST_09     Patisserie   Item_10_PAT            18.5   
1    TXN_3731986     CUST_22  Milk Products  Item_17_MILK            29.0   
2    TXN_9303719     CUST_02       Butchers   Item_12_BUT            21.5   
3    TXN_9458126     CUST_06      Beverages   Item_16_BEV            27.5   
4    TXN_4575373     CUST_05           Food   Item_6_FOOD            12.5   

   Quantity  Total Spent  Payment Method Location Transaction Date  \
0      10.0        185.0  Digital Wallet   Online       2024-04-08   
1       9.0        261.0  Digital Wallet   Online       2023-07-23   
2       2.0         43.0     Credit Card   Online       2022-10-05   
3       9.0        247.5     Credit Card   Online       2022-05-07   
4       7.0         87.5  Digital Wallet   Online       2022-10-02   

  Discount Applied  
0             True  
1             True  
2            False  
3           Unkown  
4            False

The dataset has been cleaned and is ready to be used for further analysis.